pkg/virtctl/usbredir/client.go

```diff
@@ -0,0 +1,188 @@
+		inputReader:   inReader,
+		inputWriter:   inWriter,
+		outputReader:  outReader,
+		outputWriter:  outWriter,
+		ClientConnect: clientConnect,
+	}
+}
+
+func (k *Client) WithRemoteVMIStream(usbredirStream kubecli.StreamInterface) *Client {
+	k.stream = make(chan error)
+
+	go func() {
+		defer k.outputWriter.Close()
+		k.stream <- usbredirStream.Stream(
+			kvcorev1.StreamOptions{
+				In:  k.inputReader,
+				Out: k.outputWriter,
+			},
+		)
+	}()
+
+	return k
+}
+
+func (k *Client) WithLocalTCPClient(address string) *Client {
+	lnAddr, err := net.ResolveTCPAddr("tcp", address)
+	if err != nil {
+		log.Log.Errorf("Can't resolve the address: %s", err.Error())
+		return nil
+	}
+
+	// The local tcp server is used to proxy between remote websocket and local USB
+	k.listener, err = net.ListenTCP("tcp", lnAddr)
+	if err != nil {
+		log.Log.Errorf("Can't listen on unix socket: %s", err.Error())
+		return nil
+	}
+
+	return k
+}
+
+func (k *Client) ConnectRemote() {
+	// forward data to/from websocket after usbredir client connects.
+	k.done = make(chan struct{}, 1)
+	k.remote = make(chan error)
+	go func() {
+		defer k.inputWriter.Close()
+		start := time.Now()
+
+		usbredirConn, err := k.listener.Accept()
+		if err != nil {
+			log.Log.V(2).Infof("Failed to accept connection: %s", err.Error())
+			k.remote <- err
+			return
+		}
+		defer usbredirConn.Close()
+
+		log.Log.V(2).Infof("Connected to %s at %v", usbredirClient, time.Now().Sub(start))
+
+		stream := make(chan error)
+		// write to local usbredir from pipeOutReader
+		go func() {
+			_, err := io.Copy(usbredirConn, k.outputReader)
+			stream <- err
+		}()
+
+		// read from local usbredir towards pipeInWriter
+		go func() {
+			_, err := io.Copy(k.inputWriter, usbredirConn)
+			stream <- err
+		}()
+
+		select {
+		case <-k.done: // Wait for local usbredir to complete
+		case err = <-stream: // Wait for remote connection to close
+			if err == nil {
+				// Remote connection closed, report this as error
+				err = fmt.Errorf("Remote connection has closed.")
+			}
+		}
+
+		// Wait for local usbredir to complete
+		k.remote <- err
+	}()
+}
+
+func clientConnect(device, address string) error {
+	bin := usbredirClient
+	args := []string{}
+	args = append(args, "--device", device, "--to", address)
+
+	log.Log.Infof("port_arg: '%s'", address)
+	log.Log.Infof("args: '%v'", args)
+	log.Log.Infof("Executing commandline: '%s %v'", bin, args)
+
+	command := exec.Command(bin, args...)
+	output, err := command.CombinedOutput()
+	if err != nil {
+		log.Log.Errorf("Failed to execute %v due %v, output: %v", bin, err, string(output))
+	} else {
+		log.Log.V(2).Infof("%v output: %v", bin, string(output))
+	}
+	return err
+}
+
+func (k *Client) ConnectLocal(device string) {
+	// execute local usbredir binary
+	address := k.listener.Addr().String()
+	k.local = make(chan error)
+	go func() {
+		defer close(k.done)
+		k.local <- k.ClientConnect(device, address)
+	}()
+}
+
+func (k *Client) Run() error {
```

Comment:
Run seems misleading as nothing is being ran by this function as the proxy is set up already. Maybe name this `Wait`?

Reply:
That's true. I'll change it, thanks!

Reply:
One more point is that if we decide do have `Run` or different function that will trigger the calls then it would be good if that function allows to pass context which can be used instead of `Stop`. Imho this is more common and allows you to chain contexts.

Reply:
Ack, I'll use external context then. Thanks for the review!

Reply:
I actually liked using context here and did some refactor with it. Reduced some channels and I fells better overall, thanks again for the suggestion and let me know if something feels off.